### DISCUSSION TITLE
Cropping 3D nifti file using Compose()

### DISCUSSION THREAD

**Discussion by jl3282** [Project-MONAI/MONAI#5446] [Q&A] [answered]
Hi,

I’m currently working on a project that deals with 3D nifti (nii.gz) files: one image and one label file for each patient case. Each image and label file is composed of the same shape.

I’m using “compose” from Monai to pre-process the images before performing deep learning, and I want to crop the image file to [128, 128, 128] based on the label file that indicates the targeted location. The label file consists of 1 and 0 (1 being the targeted location), and I want to find the center of the target and crop the image based on the target location. 

I initially tried to use “Resized” but later I realized that this code crops the image by zooming in when the targeted location is on the verge of the image (when there is not enough space to shape [128, 128, 128]). I want to add zero padding instead of zooming in to fit the image to the desired shape, so I found “center_of_mass” from scipy.ndimage. I want to use “center_of_mass” instead of “Resized”, but I’m not sure how I can use this in the below code.

Please help me out if anyone has faced or seen a similar problem with me. Thank you.

import torch
from monai.transforms import (
    Compose,
    LoadImaged,
    ToTensord,
    AddChanneld,
    Spacingd,
    ScaleIntensityRanged,
    CropForegroundd,
    SpatialCropd,
    Resized,
)
from monai.data import Dataset, DataLoader
from scipy.ndimage import measurements, center_of_mass, binary_dilation, zoom
train_transforms = Compose(
    [
        LoadImaged(keys=['image', 'label']),
        AddChanneld(keys=['image', 'label']),
        Spacingd(keys=['image', 'label'], pixdim=(1, 1, 1)),
        SpatialCropd(keys=['image','label'],roi_center=center_of_mass('label'==1), roi_size=[128,128,128]),
        ToTensord(keys=['image', 'label'])
    ]
)

**Comment by KumoLiu** [comment, marked as answer]
Hi @jl3282, if you want to crop the image based on the label, you could simply use `CropForeground`,  and use `margin` arg to add some margin to the crop area, then use `ResizeWithPadOrCrop` to make all images the same size. Here I give you an example.
```
CropForegroundd(keys=['image', 'label'], source_key='label', margin=[30, 30, 30]),
ResizeWithPadOrCropd(keys=['image', 'label'], spatial_size=(128, 128, 128))
```
Hope it can help you, thanks!

**Reply by jl3282** [reply]
Hi @KumoLiu , I just edit/run the code based on your advice, and it worked just as I wanted. Thank you so much!
